scipy/optimize/_milp.py

```diff
@@ -0,0 +1,280 @@
+    .. [1] Huangfu, Q., Galabova, I., Feldmeier, M., and Hall, J. A. J.
+           "HiGHS - high performance software for linear optimization."
+           Accessed 12/25/2021 at https://www.maths.ed.ac.uk/hall/HiGHS/#guide
+    .. [2] Huangfu, Q. and Hall, J. A. J. "Parallelizing the dual revised
+           simplex method." Mathematical Programming Computation, 10 (1),
+           119-142, 2018. DOI: 10.1007/s12532-017-0130-5
+
+    Examples
+    --------
+    Consider the problem at
+    https://en.wikipedia.org/wiki/Integer_programming#Example, which is
+    expressed as a maximization problem of two variables. Since `milp` requires
+    that the problem be expressed as a minimization problem, the objective
+    function coefficients on the decision variables are:
+
+    >>> c = -np.array([0, 1])
+
+    Note the negative sign: we maximize the original objective function
+    by minimizing the negative of the objective function.
+
+    We collect the coefficients of the constraints into arrays like:
+
+    >>> A = np.array([[-1, 1], [3, 2], [2, 3]])
+    >>> b_u = np.array([1, 12, 12])
+    >>> b_l = np.full_like(b_u, -np.inf)
+
+    Because there is no lower limit on these constraints, we have defined a
+    variable ``b_l`` full of values representing negative infinity. This may
+    be unfamiliar to users of `scipy.optimize.linprog`, which only accepts
+    "less than" (or "upper bound") inequality constraints of the form
+    ``A_ub @ x <= b_u``. By accepting both ``b_l`` and ``b_u`` of constraints
+    ``b_l <= A_ub @ x <= b_u``, `milp` makes it easy to specify "greater than"
+    inequality constraints, "less than" inequality constraints, and equality
+    constraints concisely.
+
+    These arrays are collected into a single `LinearConstraint` object like:
+
+    >>> from scipy.optimize import LinearConstraint
+    >>> constraints = LinearConstraint(A, b_l, b_u)
+
+    The non-negativity bounds on the decision variables are enforced by
+    default, so we do not need to provide an argument for `bounds`.
+
+    Finally, the problem states that both decision variables must be integers:
+
+    >>> integrality = np.ones_like(c)
+
+    We solve the problem like:
+    >>> from scipy.optimize import milp
+    >>> res = milp(c=c, constraints=constraints, integrality=integrality)
+    >>> res.x
+    [1.0, 2.0]
+
+    Note that had we solved the relaxed problem (without integrality
+    constraints):
+    >>> res = milp(c=c, constraints=constraints)  # OR:
+    >>> # from scipy.optimize import linprog; res = linprog(c, A, b_u)
+    >>> res.x
+    [1.8, 2.8]
+
+    we would not have obtained the correct solution by rounding to the nearest
+    integers.
+
+    """
+
+    args_iv = _milp_iv(c, integrality, bounds, constraints, options)
+    c, integrality, lb, ub, indptr, indices, data, b_l, b_u, options = args_iv
+
+    res = _highs_wrapper(c, indptr, indices, data, b_l, b_u,
+                         lb, ub, integrality, options)
```

Comment:
When I run locally, it appears that when `disp` is not explicitly set to `False`, HiGHS has logging to console set on by default, so we may need to add that option here (or in `_highs_wrapper`)

Reply:
I had code that set all the options to default values, but then I deleted it. There seemed to be some options input validation in `_highs_wrapper`, so I figured all that default-setting and such was happening there. If not, LMK where you think all of that should go. Probably should be in one place rather than split between the two.

Reply:
I added a default `'log_to_console': False`. Does `disp` map perfectly to `log_to_console` now? When we were wrapping HiGHS originally, there were a ton of different verbosity options. It looks like that may have changed?

Reply:
> Does `disp` map perfectly to `log_to_console` now?

For our purposes I think the answer is yes.

> It looks like that may have changed?

Correct, `message_level` went away altogether and there is now a separate dev logging channel in addition to standard information in `log_to_console`.  I have been advised to ignore the dev logging and stick to just `log_to_console`.